### PR TITLE
Added option for sharing achievements and a back button in ProfileActivity 

### LINKED ISSUE #4312: ProfileActivity should have a back button
**Summary:** 

There should be a back button on the top of ProfileActivity like in Settings and Review Activity.

**Steps to reproduce:** 

* Open app and tap on Hamburger Menu
* Click and open the profile.
* You can see there is no back button
I think ProfileActivity should have a back button like other Activities Like Settings and share

**Device and Android version:** 

Device: Pixel 3a
API: 27
 
**Commons app version:** 

3.0.0-debug-master~d0f895416

**Screen-shots:** 

Profile Activity: Unlike all other activities this activity doesn't have any back button

**Would you like to work on the issue?**

Yes, I would like to work on this issue. @nicolas-raoul Please assign me.

### PR DESCRIPTION
**Pull request by Ayan-10**
**Description**

Added option for sharing achievements and a back button in ProfileActivity. Also added username at the top of the toolbar

Fixes #4312 

What changes did you make and why?

There is no option in the current app for sharing achievements and it also doesn't have a back button in Profile Activity. So I added those things.

**Tests performed (required)**

Tested latest master on Pixel with API level 30.

### CODE REVIEW

**Comment by madhurgupta10** [comment]
@Ayan-10 Could you please add tests for all the new lines you added to ProfileActivity?

**Comment by Ayan-10** [comment]
@madhurgupta10 I am not good at writing tests and I recently started learning testing. But I will definitely give it a try.

**Comment by Ayan-10** [comment]
Also, @madhurgupta10 If you can suggest me some good resources for learning testing. Please tell me. 

Thanks.

**Comment by madhurgupta10** [comment]
@Ayan-10 You can look at the unit tests that are written for other methods, the best way to learn is by doing :)

**Comment by Ayan-10** [comment]
@madhurgupta10 Thank you so much for your advice. I have written the tests. Maybe there are some mistakes. Please let me know about it. I am open to learn.

**Comment by madhurgupta10** [comment]
@Ayan-10 Can you post a screenshot of the code coverage report for ProfileActivity and AchievementsFragment where we can see total lines hits/miss?

Feel free to ask any questions you have :)

**Comment by Ayan-10** [comment]
Yes, I can. But I faced some problems while running tests, first time `Command line is too long. Shorten command line for ReviewActivityTest or also for Android JUnit default configuration`  this error occurred then I fixed it by choosing to shorten command line : **JAR manifest**. But then whenever I am running any test case in the project, it's running forever and consuming lots of data. Can you help me with that?

**Comment by madhurgupta10** [comment]
Try changing your JRE from default to 1.8
Screenshot 2021-07-06 122132
Screenshot 2021-07-06 122106

**Comment by madhurgupta10** [comment]
Also set your Shorten command line to classpath file

Screenshot 2021-07-06 122448

**Comment by Ayan-10** [comment]
@madhurgupta10 I tried your suggestion setting JRE to 1.8 and shorten command line to classpath. But still, it's doing the same.

**Comment by madhurgupta10** [comment]
@Ayan-10 Unit tests can run without a virtual or a physical device, try running a single test and see if works or not. Also, post a screenshot of your test configuration.

**Comment by Ayan-10** [comment]
@madhurgupta10 This kind of client tests are running as expected.



But when It comes to test of any activity or fragment files. It's running infinitely and consuming lots of data. 




This is the configuration.

**Comment by madhurgupta10** [comment]
@Ayan-10 Your configurations look good to me, what version of Gradle Plugin are you using?

**Comment by Ayan-10** [comment]
`        classpath 'com.android.tools.build:gradle:4.0.1'`

`distributionUrl=https\://services.gradle.org/distributions/gradle-6.2.2-all.zip`

I didn't change anything in Gradle. It's the same as master.

**Comment by madhurgupta10** [comment]
@Ayan-10 Plugin version seems good to me, not sure what is causing the issue.

Not sure if these would help but few things you can try
1. Try with Looper Mode Paused
2. Increase more RAM allocation for Android Studio, close any other applications, browser running
3. Invalidate Caches and Retry
4. Delete .gradle folder, Clean and Rebuild the project

**Comment by Ayan-10** [comment]
@madhurgupta10 Tried everything but nothing worked. I am telling you the way I am running the tests. Maybe there will be some problem. 

1. Going to ProfileActivityTest file
2. Clicking the run button besides `class ProfileActivityTest {`
3. Then clicking Run ProfileActivityTest

Is it ok?

**Comment by Ayan-10** [comment]
@madhurgupta10 I did some experiments with tests and find out that only the files which are running with `RobolectricTestRunner`, are causing that problem. The tests without `@RunWith(RobolectricTestRunner::class)` this line, running as expected.

Can you help me with this fact?

**Comment by Ayan-10** [comment]
@madhurgupta10 I fixed the issue. Thanks :)

Here is the screenshot of code coverage. Please let me know about any mistakes.

**Comment by madhurgupta10** [comment]
> @madhurgupta10 I fixed the issue. Thanks :)
> 
> Here is the screenshot of code coverage. Please let me know about any mistakes.
> 
> 

Very good @Ayan-10 what was the solution btw?

For code coverage, can you post a screenshot of the ProfileActivity and AchievementsFragment instead of the package where we can see lines hits (highlighted in green) and miss (highlighted in red)

Lines hit means these lines are being covered in unit tests where miss means the opposite. When we write tests we make sure we cover max lines.

**Comment by Ayan-10** [comment]
@madhurgupta10  Actually it's an issue of robolectric. When I found the problem is with robolectric, I started learning it and while doing so I find out robolectric download some jars before running the tests. I also went to Github repo of robolectric and I found an issue regarding this problem. Downloading jars are taking a lot of time. The issue is open now they are working on it to reduce the time. And There is also an issue with android studio. When I run tests it's showing "tests are running" but that's totally misleading. At that time the jars are downloading.

**Comment by Ayan-10** [comment]
@madhurgupta10 I posted all the screenshots of ProfileActivity Tests. And I didn't add anything in achievementsFragment, I just removed some variables and methods. That's why I didn't post achievementsFragment Tests. Should I post the screenshots?

**Comment by madhurgupta10** [comment]
@Ayan-10 thanks for posting the screenshots. 91.7% is good coverage but we can still make it to 100% easily.

For the method `onOptionsItemSelected` only the default case is covered, can you add a test for the case where the `if` statement is true?

**Comment by Ayan-10** [comment]
@madhurgupta10 I added the cases for if statement true. But I am getting a `NullPointerException` on line 143 on the attached screenshot. The method on `getWindow().getDecorView().findViewById(android.R.id.content);` is giving null as a rootView. Can you help me with that?



```
    @Test
    @Throws(Exception::class)
    fun testOnOptionsItemSelected() {
        val menuItem: MenuItem = RoboMenuItem(R.menu.menu_about)
        Shadows.shadowOf(Looper.getMainLooper()).idle()
        activity.onOptionsItemSelected(menuItem)
        val menuItemShare: MenuItem = RoboMenuItem(R.id.share_app_icon)
        activity.onOptionsItemSelected(menuItemShare)
    }
```

This is the code of OnOptionsItemSelected() true case state.

**Comment by Ayan-10** [comment]
@madhurgupta10 Sorry Actually `Utils.getScreenShot(rootView);` this line is invoking NullPointerException and internally 
`Bitmap.createBitmap(screenView.getDrawingCache());` this method creating the exception. Only 2 lines are left to cover.


```
java.lang.NullPointerException
	at org.robolectric.internal.bytecode.ShadowImpl.extract(ShadowImpl.java:14)
	at org.robolectric.shadow.api.Shadow.extract(Shadow.java:25)
	at org.robolectric.shadows.ShadowBitmap.createBitmap(ShadowBitmap.java:230)
	at android.graphics.Bitmap.createBitmap(Bitmap.java)
	at fr.free.nrw.commons.Utils.getScreenShot(Utils.java:181)
	at fr.free.nrw.commons.profile.ProfileActivity.onOptionsItemSelected(ProfileActivity.java:144)
	at fr.free.nrw.commons.profile.ProfileActivityTest.testOnOptionsItemSelected(ProfileActivityTest.kt:70)
```

**Comment by Ayan-10** [comment]
Solved the issue. 97% code coverage. Only one exception coverage left.

**Comment by madhurgupta10** [comment]
@Ayan-10 Very good job. 

For Exception Test, you can use `when-thenThrow` from Mockito

File: app/src/test/kotlin/fr/free/nrw/commons/profile/ProfileActivityTest.kt [hunk 1]
```diff
@@ -65,6 +65,8 @@ class ProfileActivityTest {
         val menuItem: MenuItem = RoboMenuItem(R.menu.menu_about)
         Shadows.shadowOf(Looper.getMainLooper()).idle()
         activity.onOptionsItemSelected(menuItem)
+        val menuItemShare: MenuItem = RoboMenuItem(R.id.share_app_icon)
```
Comment (madhurgupta10) [inline app/src/test/kotlin/fr/free/nrw/commons/profile/ProfileActivityTest.kt:68]: It is a good idea to create separate test functions for each scenario

**Review by madhurgupta10 — CHANGES_REQUESTED** [review]

**Comment by Ayan-10** [comment]
@madhurgupta10 I am doing the same for throwing IOException. But every time I'm getting this 

```
org.mockito.exceptions.base.MockitoException: 
Checked exception is invalid for this method!
Invalid: java.io.IOException: IOException
```

I am trying to resolve it.

**Comment by Ayan-10** [comment]
@madhurgupta10 Mockito is not throwing IOException. I tried a lot but I can't find any way to throw IOException. It's throwing some other exceptions but not IOException. I am attaching the exception code. Can you please help me with this? I attached the Exception in the previous comment.

```
    @Test
    @Throws(Exception::class)
    fun testShareScreenException() {
        Shadows.shadowOf(Looper.getMainLooper()).idle()
        `when`(fileOutputStream.flush()).thenThrow(IOException("IOException"))
        val method: Method = ProfileActivity::class.java.getDeclaredMethod(
            "shareScreen"
        )
        method.isAccessible = true
        method.invoke(activity, bitmap)
    }

```

**Comment by madhurgupta10** [comment]
> @madhurgupta10 Mockito is not throwing IOException. I tried a lot but I can't find any way to throw IOException. It's throwing some other exceptions but not IOException. I am attaching the exception code. Can you please help me with this? I attached the Exception in the previous comment.
> 
> ```
>     @Test
>     @Throws(Exception::class)
>     fun testShareScreenException() {
>         Shadows.shadowOf(Looper.getMainLooper()).idle()
>         `when`(fileOutputStream.flush()).thenThrow(IOException("IOException"))
>         val method: Method = ProfileActivity::class.java.getDeclaredMethod(
>             "shareScreen"
>         )
>         method.isAccessible = true
>         method.invoke(activity, bitmap)
>     }
> ```

Your code looks good to me, not sure what could be the reason here. Anyways the current code coverage seems good. I am happy to merge this PR once you do the small change requested before :)

**Review by madhurgupta10 — APPROVED** [review]

**Pull request merged**